### PR TITLE
Partition support

### PR DESCRIPTION
**Pull request by Alexlandeau**
[ch58568](https://app.clubhouse.io/dataiku/story/58568/partitioning-support)

### CODE REVIEW

**Comment by alexcombessie** [comment]
Would be great to ensure we converge on a single implementation with [this PR](https://github.com/dataiku/dss-plugin-timeseries-forecast/pull/17)

**Comment by StanislasGuinel** [comment]
Once the code is ready, we'll also put it in https://github.com/dataiku/dss-plugin-dkulib to copy it from there.

**Comment by alexcombessie** [comment]
All integration tests now pass on DSS 7, thanks to your latest commit @Alexlandeau 👍 

However, I have noticed a small issue when a user adds/remove subcharts. The previously exported wordclouds are kept in the folder, which is confusing. I would recommend adding a folder clearing step in `recipe.py`, with something like `Folder.clear_partition(output_partition_path)` (pseudo-code), before writing the new wordclouds.

File: custom-recipes/nlp-visualization-wordcloud/recipe.py [hunk 1]
```diff
@@ -31,12 +32,15 @@
 # Prepare data and count tokens for each subchart
 frequencies = worcloud_visualizer.tokenize_and_count(df)
 
+# Clear output folder's target partition
+output_folder.clear_path(output_partition_path)
```
Comment (alexcombessie) [inline custom-recipes/nlp-visualization-wordcloud/recipe.py:36]: The `clear_path` method is deprecated: https://doc.dataiku.com/dss/latest/python-api/managed_folders.html#dataiku.Folder.clear_path

I would recommend using https://doc.dataiku.com/dss/latest/python-api/managed_folders.html#dataiku.Folder.delete_path or https://doc.dataiku.com/dss/latest/python-api/managed_folders.html#dataiku.Folder.clear_partition

Comment (Alexlandeau) [reply custom-recipes/nlp-visualization-wordcloud/recipe.py]: corrected sorry

File: python-lib/plugin_partitioning_utils.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    for dimension_placeholder in path_replacement_dict.keys():
+        dst_file_path = dst_file_path.replace(dimension_placeholder, path_replacement_dict[dimension_placeholder])
+
+    dst_file_path = "".join(dst_file_path.rsplit(".*"))
+
+    return dst_file_path
+
+
+@time_logging(log_message="Getting input partitions paths")
+def get_input_partitions_paths(folder_id: AnyStr) -> List[AnyStr]:
```
Comment (StanislasGuinel) [inline python-lib/plugin_partitioning_utils.py:63]: you can merge the 2 functions get_input_partitions_paths and get_output_partition_path and make one with an argument like is_input=True/False
Also, I'm not sure to understand why get_input_partitions_paths returns a list of strings and the other just a string ?

File: python-lib/plugin_partitioning_utils.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    discrete_dimensions = [dimension.get("name") for dimension in dimensions if dimension.get("type") != "time"]
+
+    # Generate path placeholders replacements dictionnary
+    path_replacement_dict = DEFAULT_OUTPUT_PATH_REPLACEMENT_DICT.copy()
+    for dimension in discrete_dimensions:
+        replacement_key = f"%{{{dimension}}}"
+        # Get flow variable containing dimension's value
+        replacement_value = dataiku.get_flow_variables()["DKU_DST_" + dimension]
+        path_replacement_dict[replacement_key] = replacement_value
+
```
Comment (StanislasGuinel) [inline python-lib/plugin_partitioning_utils.py:51]: you can use a function for complex element like this one

File: python-lib/plugin_partitioning_utils.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    # Get discrete partitioning dimensions
+    discrete_dimensions = [dimension.get("name") for dimension in dimensions if dimension.get("type") != "time"]
+
+    # Get all necessary values per partitioning dimension
+    dimension_names = [dimension.get("name") for dimension in dimensions]
+    values_per_dimension = {
+        dimension_name: dataiku.get_flow_variables()[f"DKU_SRC_{dimension_name}_VALUES"].replace("'", "").split(", ")
+        for dimension_name in dimension_names
+    }
```
Comment (StanislasGuinel) [inline python-lib/plugin_partitioning_utils.py:102]: maybe you can fusion the three loop to do only one pass over dimensions and avoid using `except IndexError`.
and also use a function to process each dimension

File: python-lib/plugin_partitioning_utils.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+            replacement_key = f"%{{{dimension}}}"
+            # Get flow variable containing dimension's value
+            replacement_value = combination.get(dimension)[0]
+            path_replacement_dict[replacement_key] = replacement_value
+
+        # Generate path
+        src_file_path = file_path_pattern
+        for dimension_placeholder in path_replacement_dict.keys():
+            src_file_path = src_file_path.replace(dimension_placeholder, path_replacement_dict[dimension_placeholder])
+            src_file_path = "".join(src_file_path.rsplit(".*"))
```
Comment (StanislasGuinel) [inline python-lib/plugin_partitioning_utils.py:131]: all this could be a function that returns a file_path

**Review by StanislasGuinel — COMMENTED** [review]
Just a few comments:
- you could merge the 2 functions into one (they share both some code)
- you could make more small functions so that there is no big function that does many different things

**Review by alexcombessie — APPROVED** [review]

**Pull request merged**